src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,421 @@
+#![allow(unused)]
+
+use alloc::{borrow::ToOwned, ffi::CString, format, string::String};
+use core::ffi::CStr;
+use ttf_parser::{GlyphId, Tag};
+use wasmtime::{self, AsContext, AsContextMut, Caller, Engine, Linker, Memory, Module, Store};
+
+use super::{
+    buffer::{hb_buffer_t, GlyphBuffer, GlyphPosition, UnicodeBuffer},
+    face::hb_glyph_extents_t,
+    hb_font_t, hb_glyph_info_t,
+    ot_shape::{hb_ot_shape_context_t, shape_internal},
+    ot_shape_plan::hb_ot_shape_plan_t,
+};
+use crate::shape_with_plan;
+
+struct ShapingData<'a> {
+    font: &'a hb_font_t<'a>,
+    buffer: hb_buffer_t,
+}
+
+pub(crate) fn shape_with_wasm(
+    // the font
+    face: &hb_font_t,
+    // the plan is useful maybe?
+    plan: &hb_ot_shape_plan_t,
+    // the text
+    mut unicode_buffer: UnicodeBuffer,
+) -> Option<GlyphBuffer> {
+    unicode_buffer.0.guess_segment_properties();
+
+    // If font has no Wasm blob just return None to carry on as usual.
+    let wasm_blob = face
+        .raw_face()
+        .table(ttf_parser::Tag::from_bytes(b"Wasm"))?;
+
+    // wasmtime stuff here
+
+    let data = ShapingData {
+        font: face,
+        buffer: unicode_buffer.0,
+    };
+
+    let mut store = Store::new(&Engine::default(), data);
+    let module = Module::new(&store.engine(), wasm_blob).ok()?; // returns None if couldn't parse blob.
+    let module_name = module.name().unwrap_or_default();
+
+    let mut linker = Linker::new(&store.engine());
+
+    // ====
+    // below are the functions that are imported by harfbuzz-wasm crate.
+    // I copied the signatures here and I am trying to satisfy the compiler as much as I can
+    // pointer land is beyond me.
+
+    // fn face_get_upem(face: u32) -> u32;
+    // Returns the units-per-em of the font face.
+    let face_get_upem = |caller: Caller<'_, ShapingData>, _face: u32| -> u32 {
+        caller.data().font.units_per_em as u32
+    };
+    linker
+        .func_wrap(module_name, "face_get_upem", face_get_upem)
+        .ok()?;
+
+    // fn font_get_face(font: u32) -> u32;
+    // Creates a new face token from the given font token.
+    let font_get_face = |caller: Caller<'_, ShapingData>, _: u32| {
+        // From HarfBuzz docs:
+        // (In the following functions, a font is a specific instantiation of a face at a
+        // particular scale factor and variation position.)
+        //
+        // I am unsure how to represent that in rustybuzz.
+        // er .. do what here?
+        0
```

Comment:
What harfbuzz does in this case?

Reply:
C++ code is inscrutable to me so I don't know. From the hafbuzz-wasm crate it seems to be just a token being passed to the host (not even a pointer).

Reply:
[that's the header file](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-wasm-api.h). 

I *think* [that's the function in question](https://github.com/harfbuzz/harfbuzz/blob/e079dd20304bacd8d753eee2ceacbaa1a3978a34/src/hb-wasm-api-font.hh#L47).

Reply:
Hm... We do not have a concept of a Font in rustybuzz/ttf-parser. Everything is a Face.
But it returns an index of some sort and not a "face". No idea what this index is suppose to represent.
Better as harfbuzz authors about it.

Reply:
If there's no Font in RB, you can just pass the Face as the Font...

The index is what represents the object inside the interpretter. You need a mapper for mapping back and force between pointers and the indices. In HB we use the OBJ2REF and REF2OBJ from our wasm runtime.